core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+	SO: SyncOracle + Send + Clone,
+	DigestItemFor<B>: CompatibleDigestItem + DigestItem<AuthorityId=Public>,
+	Error: std::error::Error + Send + From<::consensus_common::Error> + From<I::Error> + 'static,
+{
+	type OnSlot = Box<Future<Item=(), Error=consensus_common::Error> + Send>;
+
+	fn on_start(
+		&self,
+		slot_duration: u64
+	) -> Result<(), consensus_common::Error> {
+		register_babe_inherent_data_provider(&self.inherent_data_providers, slot_duration)
+	}
+
+	fn on_slot(
+		&self,
+		chain_head: B::Header,
+		slot_info: SlotInfo,
+	) -> Self::OnSlot {
+		let pair = self.local_key.clone();
+		let _public_key = self.local_key.public();
+		let client = self.client.clone();
+		let block_import = self.block_import.clone();
+		let env = self.env.clone();
+
+		let (timestamp, slot_num, slot_duration) =
+			(slot_info.timestamp, slot_info.number, slot_info.duration);
+
+		let authorities = match authorities(client.as_ref(), &BlockId::Hash(chain_head.hash())) {
+			Ok(authorities) => authorities,
+			Err(e) => {
+				error!(
+					target: "babe",
+					"Unable to fetch authorities at block {:?}: {:?}",
+					chain_head.hash(),
+					e
+				);
+				telemetry!(CONSENSUS_WARN; "babe.unable_fetching_authorities";
+					"slot" => ?chain_head.hash(), "err" => ?e
+				);
+				return Box::new(future::ok(()));
+			}
+		};
+
+		if !self.force_authoring && self.sync_oracle.is_offline() && authorities.len() > 1 {
+			debug!(target: "babe", "Skipping proposal slot. Waiting for the network.");
+			telemetry!(CONSENSUS_DEBUG; "babe.skipping_proposal_slot";
+				"authorities_len" => authorities.len()
+			);
+			return Box::new(future::ok(()));
+		}
+
+		// FIXME replace the dummy empty slices with real data
+		let (proposal_work, vrf_output, proof) = if let Some((inout, proof, _batchable_proof)) = author_block(
+			&[0u8; 0],
+			slot_info.number,
+			&[0u8; 0],
+			0,
+			&authorities,
+			&pair,
+			self.threshold,
+		) {
+			debug!(
+				target: "babe", "Starting authorship at slot {}; timestamp = {}",
+				slot_num,
+				timestamp,
+			);
+			telemetry!(CONSENSUS_DEBUG; "babe.starting_authorship";
+				"slot_num" => slot_num, "timestamp" => timestamp
+			);
+
+			// we are the slot author. make a block and sign it.
+			let proposer = match env.init(&chain_head, &authorities) {
+				Ok(p) => p,
+				Err(e) => {
+					warn!(target: "babe", "Unable to author block in slot {:?}: {:?}", slot_num, e);
+					telemetry!(CONSENSUS_WARN; "babe.unable_authoring_block";
+						"slot" => slot_num, "err" => ?e
+					);
+					return Box::new(future::ok(()))
+				}
+			};
+
+			let remaining_duration = slot_info.remaining_duration();
+			// deadline our production to approx. the end of the
+			// slot
+			(Timeout::new(
+				proposer.propose(
+					slot_info.inherent_data,
+					remaining_duration,
+				).into_future(),
+				remaining_duration,
+			),
+			inout.to_output(),
+			proof)
+		} else {
+			return Box::new(future::ok(()));
```

Comment:
I guess the same goes here.

Reply:
AuRa also has the same problem

Reply:
i think this is something to solve at the proposer level. It could be that due to a runtime error we aren't able to include transactions anymore. The proposer logic should detect when failures are occurring too often and start to issue empty blocks for some time so that the chain continues to progress.

Reply:
I agree, but that is a separate issue.